### ISSUE TITLE
Canonicalize error due to scalar product

### ISSUE THREAD

**Issue by Mathijssch** [cvxpy/cvxpy#1358]
**Describe the bug**
I am trying to solve a rather large-scale DPP-compliant parametric problem, the code for which is too complex to paste here. I am able to successfully build the problem, but when calling the solver, I obtain a `TypeError` akin to the one pasted below, which seems to arise in the canonicalization of part of the problem. 

**To Reproduce**
After some digging I found that it seems like the root of the problem lies in canonicalizing a scalar product of a parameter and a decision variable. The following snippet yields the same error.

```
import cvxpy as cp 
p = cp.Parameter((4,))
v = cp.Variable((4,))
cp.scalar_product(p,v).canonicalize()
```

**Expected behavior**
Being able to canonicalize expressions involving the scalar product of a parameter and a decision variable. 

**Output**

The MWE outputs 
```
Traceback (most recent call last):
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/performance_utils.py", line 33, in _lazyprop
    return getattr(self, attr_name)
AttributeError: 'multiply' object has no attribute '_lazy_canonical_form'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/performance_utils.py", line 33, in _lazyprop
    return getattr(self, attr_name)
AttributeError: 'conj' object has no attribute '_lazy_canonical_form'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/atoms/atom.py", line 315, in canonicalize
    obj, constr = arg.canonical_form
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/performance_utils.py", line 35, in _lazyprop
    setattr(self, attr_name, func(self))
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/canonical.py", line 42, in canonical_form
    return self.canonicalize()
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/atoms/atom.py", line 315, in canonicalize
    obj, constr = arg.canonical_form
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/performance_utils.py", line 35, in _lazyprop
    setattr(self, attr_name, func(self))
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/utilities/canonical.py", line 42, in canonical_form
    return self.canonicalize()
  File "/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/cvxpy/atoms/atom.py", line 322, in canonicalize
    data)
TypeError: cannot unpack non-iterable NotImplementedType object
```

**Version**
 - OS: Ubuntu 20.04
 - CVXPY Version: 1.1.1 **[[EDIT]]** I have also tried with the latest version (1.1.11), but the issue persists.

**Comment by SteveDiamond** [comment]
Try updating cvxpy. You have an old version.

**Comment by SteveDiamond** [comment]
Sorry for the slow response. I looked into this and indeed there is an error with the code you posted above. I'm not sure though that the code path should ever be called in reality though. For example, you can solve a problem involving the scalar product of a parameter and a variable just fine:
```
import cvxpy as cp 
import numpy as np
p = cp.Parameter((4,))
v = cp.Variable((4,))

p.value = np.ones(4)
obj = cp.Minimize(p@v)
prob = cp.Problem(obj, [v >= 1])
prob.solve(verbose=True)
```
Anyway I'm not saying you're wrong necessarily about the source of your bug, I'd just like more information on why you think it's the source.

**Comment by Mathijssch** [comment]
Your example does indeed seem to work as it should, but it seems there is some subtlety in the `@`-operator somehow circumventing the problem. Replacing `p@v` with `cp.scalar_product(p,v)` in your example triggers the error again: 
```
import cvxpy as cp
import numpy as np
p = cp.Parameter((4,))
v = cp.Variable((4,))

p.value = np.ones((4,))
obj = cp.Minimize(cp.scalar_product(v, p))
# obj = cp.Minimize(v@p)
prob = cp.Problem(obj, [v >= 1])
prob.solve(verbose=True)
```

**Comment by rileyjmurray** [comment]
The ``scalar_product`` atom was introduced to help with more abstract vector spaces, like spaces where a "vector" is represented by list of expressions with varying shapes. Another intended use is for matrices, where cvxpy parses ``cp.scalar_product(A, X)`` much more efficiently than ``cp.trace(A.T @ X)``. In the example you give it's definitely better to use ``v @ p`` rather than ``scalar_product``.

Of course, all that said, there is clearly a bug here. When I implemented ``scalar_product`` I don't think I tested against Parameter inputs. Thanks for pointing this out @Mathijssch; we'll fix this.
